### PR TITLE
refactor(pydantic): add support for pydantic v2

### PR DESCRIPTION
**Pull request by JeroennC**
Supports v1 and v2 simultaneously, although deprecated features are used, namely: Config class, root_validator, .dict()

### CODE REVIEW

File: kstreams/backends/kafka.py [hunk 1]
```diff
@@ -54,7 +54,7 @@ class Kafka(BaseModel):
     bootstrap_servers: List[str] = ["localhost:9092"]
     security_protocol: SecurityProtocol = SecurityProtocol.PLAINTEXT
 
-    ssl_context: Optional[ssl.SSLContext]
+    ssl_context: Optional[ssl.SSLContext] = None
```
Comment (JeroennC) [inline kstreams/backends/kafka.py:57]: Optional is no longer enough to mean not required
https://docs.pydantic.dev/latest/migration/#required-optional-and-nullable-fields

File: kstreams/backends/kafka.py [hunk 2]
```diff
@@ -65,7 +65,7 @@ class Config:
         arbitrary_types_allowed = True
         use_enum_values = True
 
-    @root_validator
+    @root_validator(skip_on_failure=True)
```
Comment (JeroennC) [inline kstreams/backends/kafka.py:68]: skip_on_failure must be true for pydantic v2. In this case it does not matter as this validator does not need to run if other validations already failed.

https://docs.pydantic.dev/latest/migration/#validator-and-root_validator-are-deprecated

File: tests/test_backend_kafka.py [hunk 3]
```diff
@@ -50,7 +50,7 @@ def test_sasl_plain_fail_missing_pass():
             security_protocol=SecurityProtocol.SASL_PLAINTEXT,
             sasl_plain_username=username,
         )
-    assert "sasl_plain_password" in str(e.value.args[0])
```
Comment (JeroennC) [inline tests/test_backend_kafka.py:53]: This test relied on the internals of ValidationError, which changed

File: kstreams/backends/kafka.py (hunk 2 above)
Comment (marcosschroh) [reply kstreams/backends/kafka.py:68]: Shouldn't it be replaced by [model_validator](https://docs.pydantic.dev/latest/api/functional_validators/#pydantic.functional_validators.model_validator)?

Comment (JeroennC) [reply kstreams/backends/kafka.py:68]: For pydantic v2, yes - but in this PR i am maintaining support for pydantic v1 so as not to be a breaking change

**Review by marcosschroh — APPROVED** [review]

**Pull request merged**